Add recipe for org-latex-instant-preview

yangsheng6810 opened the pull request:
### Brief summary of what the package does

It provides instant preview of LaTeX snippets in `org-mode` via MathJax outputed SVG

### Direct link to the package repository

https://github.com/yangsheng6810/org-latex-instant-preview

### Your association with the package

I am the maintainer 

### Relevant communications with the upstream package maintainer

None needed

### Checklist



Please confirm with `x`:

- [X] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [X] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [X] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [X] My elisp byte-compiles cleanly
- [X] `M-x checkdoc` is happy with my docstrings
- [X] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

NOTE: Since `package-lint` is not compatible with `names`, a package I used for namespacing, I took the macro expanded version to run through `package-lint`. I received the following errors, which I think are all false positives.
```
11:36: error: Package names is not installable.
11:52: error: Package s is not installable.
11:64: error: Package posframe is not installable.
11:83: error: Package org is not installable.
11:95: error: Package dash is not installable.
34:4: error: Aliases should start with the package's prefix "org-latex-instant-preview".
39:0: error: "posframe-poshandler-point-window-center" doesn't start with package's prefix "org-latex-instant-preview".
```
Since package install works according to directions in CONTRIBUTING.org, I assume the first 5 errors are false positive. The error regarding alias is a workaround for a bug in `names`. The error regarding `posframe-poshandler-point-window-center` adds a function to show a posframe at some specific point. I have made a pull request regarding this function to `posframe.el`, which has been submitted and (partially) approved, but it is waiting for my FSF assignment before it can be merged. While we are waiting, I am adding a separate implementation, which will be removed once my pull request to `posframe` is merged.

riscy commented:
Looks interesting, I hadn't heard of MathJax.

Yes, package-lint will misfire on `names`.  In my humble opinion you are making a sacrifice by inviting extra noise into your package-lint signal, which will make issues easier to miss going forward.  But this is also not in my view a blocker -- thanks for the note.

Quick first pass from my end, with the package-lint output edited:

### org-latex-instant-preview.el ###

byte-compile (using Emacs 27.1):
``
org-latex-instant-preview.el:38:1:Warning: Unused lexical variable
    `mode-line-height'
org-latex-instant-preview.el:38:1:Warning: Unused lexical variable
    `posframe-height'
org-latex-instant-preview.el:38:1:Warning: Unused lexical variable
    `window-height'
org-latex-instant-preview.el:38:1:Warning: Unused lexical variable
    `window-top'
In end of data:
org-latex-instant-preview.el:466:1:Warning: the function
    `posframe-poshandler-point-window-center' is not known to be defined.
```
I think the last of these can be handled by adding a `declare-function`

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20200906.512):
```
458:8: warning: Closing parens should not be wrapped onto new lines.
```

Other possible lints:
- org-latex-instant-preview.el#L34: It's safer to sharp-quote function names; use `#'`
- org-latex-instant-preview.el#L64: Prefer `https` over `http` if possible ([why?](https://news.ycombinator.com/item?id=22933774))

### Package ###

- GitHub API found `The Unlicense`
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to org-latex-instant-preview.el

yangsheng6810 commented:
MathJax is a javascript library that renders LaTeX equations on a web page, and you only need to do exactly what you will do in LaTeX. It is behind perhaps most equations you will see online (other major options are KeTeX or rendered images). I am using it as it it is a lightweight solution for LaTeX equqation rendering.

> In my humble opinion you are making a sacrifice by inviting extra noise into your package-lint signal, which will make issues easier to miss going forward.

This is my first Emacs package, and it seems I chose a name way too long. And I cannot think of a shorter name. It will be a big headache to get rid of `names` since variables and functions will be way too long. It can also be a problem since changing the name and namespace of a package could break peoples' configurations. I am willing to get rid of `names` if these two issues can be handled, and better do it before it is available on Melpa. Do you have any suggestions for that? Maybe a different name for the package?

> I think the last of these can be handled by adding a `declare-function`

I am confused here. I assume `declare-function` tells the byte-compiler where to find a function, but this function is not in the `posframe` package yet (hopefully it will arrive soon). Will it cause any trouble if the function is not define there? Notice I am defining one when it is not defined in `posframe`.

> org-latex-instant-preview.el#L34: It's safer to sharp-quote function names; use `#'`

Fixed.

>  org-latex-instant-preview.el#L64: Prefer `https` over `http` if possible ([why?](https://news.ycombinator.com/item?id=22933774))

Fixed.

BTW, after some research and thoughts, I am switch to GPLv3 license.

yangsheng6810 commented:
@riscy I decided to get rid of the dependency of `names`, and give the package a shorter name: `org-latex-impatient`. And the undeclared function problem is handled with a `eval-and-compile`. Does it look good now?

riscy commented:
Apologies for the slow turnaround!  Looking again.

> I am confused here. I assume declare-function tells the byte-compiler where to find a function, but this function is not in the posframe package yet (hopefully it will arrive soon).

Sorry, I misunderstood -- please disregard my declare-function suggestion.

For the record, I'd recommend against anyone (even you) redefining `posframe-poshandler-point-window-center` into their package while waiting for https://github.com/tumashu/posframe/pull/78 to be merged.  That PR could get held up, and it is generally bad to have multiple definitions of the same function floating around in the Emacs ecosystem.  Is it possible to name the function something that uses your package's prefix in the meantime?  If not, I think this is minor, and this will be an easy merge after I hear back from you.

yangsheng6810 commented:
Renamed the function to be within `org-latex-impatient` namespace.

riscy commented:
Great, thanks!  And welcome to MELPA. :)
